File: forcelayout/src/main/java/jp/kai/forcelayout/Forcelayout.java
package jp.kai.forcelayout;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.graphics.RectF;
import android.util.Log;
import android.util.Pair;
import android.view.MotionEvent;
import android.view.View;
import android.view.Display;
import android.view.WindowManager;

import java.net.URI;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;


/**
 * Created by kai on 2016/09/03.
 */
public class Forcelayout extends View{
    private static Properties properties = null;
//    private static Nodestyle nodestyle = null;
//    private static Nodestyle nodestyle = null;
    private static Context mContext = null;
    private static HashMap<String, Bitmap> nodeslist = new HashMap<>();
    public static Properties.Node[] nodes = new Properties.Node[200];
    public static Properties.Edge[] edges = new Properties.Edge[500];
    static ArrayList<String> nodename_array = new ArrayList<String>();
    static ArrayList<Bitmap> nodebitmap_array = new ArrayList<>();
    private static ArrayList<String> convertlist = new ArrayList<>();;
    private static int nedges = 0;
    private static float display_width;
    private static float display_height;
    private static int nodeindex = 0; //number of nodes

    private int targetnode = -1;

    private static int roundsize = 5;
    private static int fontsize = 30;
    private static boolean drawline = true;
    private static boolean drawlabel = true;

    public Forcelayout(Context context) {
        super(context);
        mContext = context;
        init();
    }

    private static void init(){
        Properties.Node[] nodes = new Properties.Node[200];
        Properties.Edge[] edges = new Properties.Edge[500];
        nodename_array.clear();
        nodebitmap_array.clear();
        convertlist.clear();
        nedges = 0;
        nodeindex = 0;
    }

    public static Properties with(Context context){
        mContext = context;
        properties = new Properties(context);

        return properties.prepare();
    }

//    public static Nodestyle node(Context context){
//        mContext = context;
//        nodestyle = new Nodestyle(context);
//        return nodestyle.prepare();
//    }
//
//    public static Linkstyle link(Context context){
//        mContext = context;
//        linkstyle = new Linkstyle(context);
//        return linkstyle.prepare();
//    }


    @Override
    public boolean onTouchEvent(MotionEvent event) {
        int touch_x = (int)event.getX();
        int touch_y = (int)event.getY();

        switch ( event.getAction() ) {

            case MotionEvent.ACTION_DOWN:
                for(int i=0; i< nodeindex; i++){
                    if((nodes[i].x + nodes[i].width >= touch_x && nodes[i].x <= touch_x)
                            && (nodes[i].y + nodes[i].height >= touch_y && nodes[i].y <= touch_y)){
                        targetnode = i;
                    }
                }
                break;

            case MotionEvent.ACTION_MOVE:

                if(targetnode != -1){
                    nodes[targetnode].x = touch_x-nodes[targetnode].width/2;
                    nodes[targetnode].y = touch_y-nodes[targetnode].height/2;
                }
                break;

            case MotionEvent.ACTION_UP:
                targetnode = -1;
                break;

            case MotionEvent.ACTION_CANCEL:
                targetnode = -1;
                break;

        }

        return true;
    }


    // draw function
    @Override
    protected void dispatchDraw(Canvas canvas) {
        Paint paint = new Paint();

        //draw link's line
        if(edges.length != 0){
            for (int i = 0 ; i < nedges-1 ; i++) {
                if (edges[i].group) {
                    Properties.Edge e = edges[i];
                    float x1 = (float) (nodes[e.from].x + nodes[e.from].width/2);
                    float y1 = (float) (nodes[e.from].y + nodes[e.from].height/2);
                    float x2 = (float) (nodes[e.to].x + nodes[e.to].width/2);
                    float y2 = (float) (nodes[e.to].y + nodes[e.to].height/2);

                    if(drawline) {
                        paint.setStrokeWidth(5);
                        float[] pts = {x1, y1, x2, y2};
                        canvas.drawLines(pts, paint);
                    }
                }
            }
        }

        //draw node's image
        for (final String str : nodeslist.keySet()) {
            if(convertlist.indexOf(str) == -1) {
                Bitmap _bitmap = nodeslist.get(str);
                nodebitmap_array.add(getCroppedBitmap(nodeslist.get(str), roundsize));
                convertlist.add(str);
            }
        }

        //draw label
        for (int i = 0; i < convertlist.size(); i++) {
            canvas.drawBitmap(nodebitmap_array.get(i), (int) nodes[i].x, (int) nodes[i].y, paint);

            if(drawlabel) {
                paint.setTextSize(30);
                canvas.drawText(nodes[i].nodename, (int) (nodes[i].x + nodes[i].width), (int) (nodes[i].y + nodes[i].height + 30), paint);
            }
        }

        if(nedges != 0) {

            properties.relax();
        }

        invalidate();
    }


//    public static class Test <L, I> {
//        private final L label;
//        private final I image;
//
//        private Test(L label, I image){
//            this.label = label;
//            this.image = image;
//
//        }
//
//        public static <L, I> Test<L, I> create(L label, I image) {
//            return new Test<>(label, image);
//        }
//    }


    //
    public Bitmap getCroppedBitmap(Bitmap bitmap, int round) {

        int width  = bitmap.getWidth();
        int height = bitmap.getHeight();

        final Rect rect   = new Rect(0, 0, width, height);
        final RectF rectf = new RectF(0, 0, width, height);

        Bitmap output = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(output);

        final Paint paint = new Paint();
        paint.setAntiAlias(true);

        canvas.drawRoundRect(rectf, width / round, height / round, paint);
        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        canvas.drawBitmap(bitmap, rect, rect, paint);
        return output;

    }


//    public static class Nodestyle{
//        private Context mContext;
//
//        public Nodestyle(Context context){
//            mContext = context;
//        }
//
//        private Nodestyle prepare(){
//            return this;
//        }
//
//        //setter of node's size
//        public Nodestyle nodesize(int nodeswidth){
//            this.nodeswidth = nodeswidth;
//            return this;
//        }
//
//        private Nodestyle nodes(){
//            return this;
//        }
//
//
//    }



    public static class Properties{
        private static Context mContext;

        private static float nodearea_width; //draw area = screen size
        private static float nodearea_height;
        private static int reduction = 30;
        private static int nodeswidth = 150; //node's width
        private static int distance = 300; //distance between nodes


        private static double gravity = 0.04;
        private static double bounce = 0.08; //
        private static double attenuation = 0.7;//0.9; //
        private static double coulomb = 680; //


        public Properties(Context context){
            mContext = context;
        }

        private Properties prepare(){
            Display mDisplay = getDisplayMetrics(mContext);
            display_width = mDisplay.getWidth();
            display_height = mDisplay.getHeight();

            return this;
        }

        //setter of linkStrength
        public Properties linkStrength(double bounce){
            this.bounce = bounce;
            return this;
        }

        //setter of linkStrength
        public Properties distance(int distance){
            this.distance = distance;
            return this;
        }

        //setter of linkStrength
        public Properties gravity(double gravity){
            this.gravity = gravity;
            return this;
        }

        //setter of node's size
        public Properties nodesize(int nodeswidth){
            this.nodeswidth = nodeswidth;
            return this;
        }

        //draw line
        public Properties drawLine(boolean drawline){
            Forcelayout.drawline = drawline;
            return this;
        }

        //draw label
        public Properties drawLable(boolean drawlabel){
            Forcelayout.drawlabel = drawlabel;
            return this;
        }

        //set fontSize
        public Properties setFontSize(int fontsize){
            Forcelayout.fontsize = fontsize;
            return this;
        }


        //setter of nodes : drawable
        public Properties nodes(final HashMap<String, Integer> nodemaps){
            init_nodes();

            Resources resource = mContext.getResources();
            for (final String str : nodemaps.keySet()) {
                BitmapFactory.Options imageOptions = new BitmapFactory.Options();
                imageOptions.inPreferredConfig = Bitmap.Config.ARGB_8888;
                imageOptions.inJustDecodeBounds = true;
                BitmapFactory.decodeResource(resource, nodemaps.get(str), imageOptions);
                // get image width
                int bitmapwidth  = imageOptions.outWidth;

                BitmapFactory.Options bmfOptions = new BitmapFactory.Options();

                //resize
                reduction = bitmapwidth / nodeswidth;
                if(reduction != 0) {
                    bmfOptions.inSampleSize = reduction;
                }

                Bitmap bitmap = BitmapFactory.decodeResource(resource, nodemaps.get(str), bmfOptions);

                int imgheight = bmfOptions.outHeight;
                int imgwidth = bmfOptions.outWidth;


                if(imgwidth != nodeswidth){
                    bitmap = resizeBitmap(bitmap, nodeswidth);

                    imgheight = bitmap.getHeight();
                    imgwidth = bitmap.getWidth();
                }

                nodearea_width = display_width - (int) (imgwidth);
                nodearea_height = display_height - (int) (imgheight);

                addNode(str, nodeindex, imgwidth, imgheight);

                nodename_array.add(str);

                nodeslist.put(str, bitmap);
                nodeindex++;
            }

            return this;
        }


        //setter of links
        public Properties links(final List<String> linkmaps){
            init_links();

            for(int i=0; i < nodename_array.size();i++){
                for(int j=0; j < nodename_array.size(); j++){
                    if(i != j) {
                        addEdge(i, j);
                    }
                }
            }

            for (int k=0; k<linkmaps.size(); k++) {
                String[] pair = linkmaps.get(k).split("-");

                if(pair.length == 2) {
                    for (int i = 0; i < nedges; i++) {
                        if ((edges[i].from == nodename_array.indexOf(pair[0]) && edges[i].to == nodename_array.indexOf(pair[1]) ||
                                (edges[i].to == nodename_array.indexOf(pair[0]) && edges[i].from == nodename_array.indexOf(pair[1])))) {
                            edges[i].group = true;
                        }
                    }
                }
            }

            return  this;
        }

        //class of node
        public static class Node{
            String nodename;

            double x;
            double y;
            double width;
            double height;

            double dx;
            double dy;
        }

        //class of link
        public static class Edge {
            int from;
            int to;
            double len;
            boolean group;
        }


        public static void addNode(String lbl, int index, int width, int height){
            Node n = new Node();

            n.x = (nodearea_width)*Math.random();
            n.y = (nodearea_height-10)*(Math.random())+10;
            n.nodename = lbl;
            n.width = width;
            n.height = height;
            n.dx = 0;
            n.dy = 0;

            nodes[index] = n;
        }

        public static void addEdge(int from, int to){
            Edge e = new Edge();
            e.from = from;
            e.to = to;
            e.group = false;
            edges[nedges++] = e;
        }


        public void relax(){
            if(nedges != 0){
                for(int i=0; i<nodeindex; i++){
                    double fx = 0,fy = 0;

                    for(int j=0; j<nodeindex; j++){

                        double distX = (int)((nodes[i].x + nodes[i].width/2) - (nodes[j].x + nodes[j].width/2));
                        double distY = (int)((nodes[i].y + nodes[i].height/2) - (nodes[j].y + nodes[j].height/2));
                        double rsq = distX * distX + distY *distY;
                        int rsq_round = (int)rsq*100;
                        rsq = rsq_round/100;

                        double coulombdist_x = coulomb * distX;
                        double coulombdist_y = coulomb * distY;
                        int coulombdist_round_x = (int)coulombdist_x*100;
                        int coulombdist_round_y = (int)coulombdist_y*100;
                        coulombdist_x = coulombdist_round_x/100;
                        coulombdist_y = coulombdist_round_y/100;



                        if(rsq != 0 && Math.sqrt(rsq) < distance) {
                            fx += coulombdist_x / rsq ;
                            fy += coulombdist_y / rsq ;
                        }
                    }

                    //gravity : node - central
                    double distX_c=0,distY_c=0;
                    distX_c = display_width/2 - (nodes[i].x + nodes[i].width/2);
                    distY_c = display_height/2 - (nodes[i].y + nodes[i].height/2);

                    fx += gravity *distX_c;
                    fy += gravity *distY_c;


                    //node in group : from - to
                    for(int j=0; j<nedges-1; j++){
                        double distX=0,distY=0;
                        if(edges[j].group) {
                            if (i == edges[j].from) {
                                distX = nodes[edges[j].to].x - nodes[i].x;
                                distY = nodes[edges[j].to].y - nodes[i].y;

                            } else if (i == edges[j].to) {
                                distX = nodes[edges[j].from].x - nodes[i].x;
                                distY = nodes[edges[j].from].y - nodes[i].y;
                            }
                        }
                        fx += bounce *distX;
                        fy += bounce *distY;
                    }

                    nodes[i].dx = (nodes[i].dx + fx) * attenuation;
                    nodes[i].dy = (nodes[i].dy + fy) * attenuation;


                    nodes[i].x += nodes[i].dx;
                    nodes[i].y += nodes[i].dy;

                }
            }
        }

        private static void init_nodes(){
            Properties.Node[] nodes = new Properties.Node[200];
            nodename_array.clear();
            nodebitmap_array.clear();
            convertlist.clear();
            nodeindex = 0;
        }

        private static void init_links(){
            Properties.Edge[] edges = new Properties.Edge[500];
            nedges = 0;
        }


        //display size
        public static final Display getDisplayMetrics(Context context) {
            WindowManager winMan = (WindowManager)context.getSystemService(Context.WINDOW_SERVICE);
            Display disp = winMan.getDefaultDisplay();
            return disp;
        }

        public static Bitmap resizeBitmap(Bitmap src, int width){
            int srcWidth = src.getWidth();
            int srcHeight = src.getHeight();

            // get Screen size
            Matrix matrix = new Matrix();

            float widthScale = (float)width / (float)srcWidth;
            matrix.postScale(widthScale, widthScale);


            // resize
            Bitmap dst = Bitmap.createBitmap(src, 0, 0, srcWidth, srcHeight, matrix, true);
            src.recycle();
            src = null;

            return dst;
        }

    }
}

File: forcelayout/src/main/kotlin/jp/kai/forcelayout/Forcelayout.kt
package jp.kai.forcelayout

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PointF
import android.util.AttributeSet
import android.util.Pair
import android.view.MotionEvent
import android.view.ScaleGestureDetector
import android.view.View
import jp.kai.forcelayout.properties.ForceProperty
import jp.kai.forcelayout.properties.GraphStyle
import jp.kai.forcelayout.properties.GraphStyle.defaultColor
import jp.kai.forcelayout.properties.LinkProperty
import jp.kai.forcelayout.properties.NodeProperty

/**
 * Created by kai on 2017/05/01.
 * Main Class
 */

open class Forcelayout : View {
    /** instance */
    private val forceProperty: ForceProperty
    private val nodeProperty: NodeProperty = NodeProperty()
    private val linkProperty: LinkProperty = LinkProperty()

    private var targetNode = -1

    private var touchX: Float = 0f
    private var touchY: Float = 0f

    private var pointF: PointF = PointF()
    private var scaleFactor = 1f
    private var detector: ScaleGestureDetector? = null

    private var startX = 0f
    private var startY = 0f
    private var translateX = 0f
    private var translateY = 0f
    private var previousTranslateX = 0f
    private var previousTranslateY = 0f

    // TODO enumか何かで管理する: pan/expand/drag
    private var zooming: Boolean = false

    init {
        detector = ScaleGestureDetector(context, ScaleListener())
    }

    constructor(mContext: Context) : super(mContext) {
        forceProperty = ForceProperty(mContext)
    }

    constructor(mContext: Context, attrs: AttributeSet) : super(mContext, attrs) {
        forceProperty = ForceProperty(mContext)
    }

    constructor(mContext: Context, attrs: AttributeSet, defStyleAttr: Int) : super(mContext, attrs, defStyleAttr) {
        forceProperty = ForceProperty(mContext)
    }

    /**
     * Create Builders
     */
    fun node(): NodeProperty {
        return nodeProperty.prepare()
    }

    fun link(): LinkProperty {
        return linkProperty.prepare()
    }

    fun with(): ForceProperty {
        return forceProperty.prepare()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        this.touchX = event.x
        this.touchY = event.y

        when (event.action) {

            MotionEvent.ACTION_DOWN -> {

                //TODO 位置補正はタッチ時の判定でやったほうが楽そうなので、頑張る translateとscaleFactorあたりを調査

                if (targetNode == -1) {
                    for (i in 0 until forceProperty.nodeIndex) {
                        if (forceProperty.nodes[i].x + forceProperty.nodes[i].width >= (touchX - translateX) / scaleFactor &&
                                forceProperty.nodes[i].x <= (touchX - translateX) / scaleFactor &&
                                forceProperty.nodes[i].y + forceProperty.nodes[i].height >= (touchY - translateY) / scaleFactor &&
                                forceProperty.nodes[i].y <= (touchY - translateY) / scaleFactor) {

                            targetNode = i
                        }
                    }
                    if (targetNode == -1) {
                        // TODO koko?
                        startX = event.x - previousTranslateX
                        startY = event.y - previousTranslateY
                    }

                } else {
                    forceProperty.nodes[targetNode].x = ((touchX - translateX) / scaleFactor - forceProperty.nodes[targetNode].width / 2)
                    forceProperty.nodes[targetNode].y = ((touchY - translateY) / scaleFactor - forceProperty.nodes[targetNode].height / 2)

                }


            }

            MotionEvent.ACTION_MOVE -> {
                if (!zooming) {

                    if (targetNode == -1) {
                        // TODO koko?
                        translateX = event.x - startX
                        translateY = event.y - startY

                    } else {
//                        forceProperty.nodes[targetNode].x = touchX - forceProperty.nodes[targetNode].width / 2
//                        forceProperty.nodes[targetNode].y = touchY - forceProperty.nodes[targetNode].height / 2
                        forceProperty.nodes[targetNode].x = ((touchX - translateX) / scaleFactor - forceProperty.nodes[targetNode].width / 2)
                        forceProperty.nodes[targetNode].y = ((touchY - translateY) / scaleFactor - forceProperty.nodes[targetNode].height / 2)
                    }
                }
            }

            MotionEvent.ACTION_UP -> {
                zooming = false
                targetNode = -1

                previousTranslateX = translateX
                previousTranslateY = translateY
            }

            MotionEvent.ACTION_POINTER_UP -> {
                zooming = false

                previousTranslateX = translateX
                previousTranslateY = translateY
            }

            MotionEvent.ACTION_POINTER_DOWN -> {
                midPoint(pointF, event)
                zooming = true
                targetNode = -1
            }

            MotionEvent.ACTION_CANCEL -> {
                zooming = false
                targetNode = -1

                previousTranslateX = translateX
                previousTranslateY = translateY
            }

        }
        detector!!.onTouchEvent(event)

        return true
    }


    private inner class ScaleListener : ScaleGestureDetector.SimpleOnScaleGestureListener() {
        override fun onScale(detector: ScaleGestureDetector): Boolean {
            scaleFactor *= detector.scaleFactor
            scaleFactor = Math.max(MIN_ZOOM, Math.min(scaleFactor, MAX_ZOOM))
            return true
        }
    }

    public override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)

    }

    // draw function
    override fun dispatchDraw(canvas: Canvas) {
        val paint = Paint()

        // scale the canvas
        canvas.scale(scaleFactor, scaleFactor, pointF.x, pointF.y)

        canvas.translate(translateX / scaleFactor, translateY / scaleFactor)

        if (targetNode != -1) {
            forceProperty.nodes[targetNode].x = ((touchX - translateX) / scaleFactor - forceProperty.nodes[targetNode].width / 2)// / scaleFactor
            forceProperty.nodes[targetNode].y = ((touchY - translateY) / scaleFactor - forceProperty.nodes[targetNode].height / 2)// / scaleFactor
        }

        if (forceProperty.isReady) {
            //draw link's line
            for (i in 0 until forceProperty.nEdges) {
                if (forceProperty.edges[i].group) {
                    val e = forceProperty.edges[i]
                    val x1 = (forceProperty.nodes[e.from].x + forceProperty.nodes[e.from].width / 2).toFloat()
                    val y1 = (forceProperty.nodes[e.from].y + forceProperty.nodes[e.from].height / 2).toFloat()
                    val x2 = (forceProperty.nodes[e.to].x + forceProperty.nodes[e.to].width / 2).toFloat()
                    val y2 = (forceProperty.nodes[e.to].y + forceProperty.nodes[e.to].height / 2).toFloat()

                    paint.strokeWidth = GraphStyle.linkWidth
                    paint.color = GraphStyle.linkColor
                    canvas.drawLine(x1, y1, x2, y2, paint)
                }
            }
            if (GraphStyle.isImgDraw) {
                /** draw node images and labels */
                val iterator: Iterator<Pair<String, Bitmap>> = forceProperty.nodesList.iterator()
                var index = 0
                while (iterator.hasNext()) {
                    val pair: Pair<String, Bitmap> = iterator.next()

                    paint.color = defaultColor
                    canvas.drawBitmap(pair.second, forceProperty.nodes[index].x.toFloat(), forceProperty.nodes[index].y.toFloat(), paint)

                    paint.textSize = GraphStyle.fontSize
                    paint.color = GraphStyle.fontColor
                    canvas.drawText(forceProperty.nodes[index].nodename, (forceProperty.nodes[index].x + forceProperty.nodes[index].width).toFloat(), (forceProperty.nodes[index].y + forceProperty.nodes[index].height + 30.0).toFloat(), paint)

                    index++
                }
            } else {
                paint.color = GraphStyle.nodeColor

                for (i in 0 until forceProperty.nodeIndex) {
                    paint.color = GraphStyle.nodeColor
                    canvas.drawCircle((forceProperty.nodes[i].x + GraphStyle.nodesWidth / 2).toFloat(), (forceProperty.nodes[i].y + GraphStyle.nodesWidth / 2).toFloat(), (GraphStyle.nodesWidth / 2).toFloat(), paint)

                    paint.textSize = GraphStyle.fontSize
                    paint.color = GraphStyle.fontColor
                    canvas.drawText(forceProperty.nodes[i].nodename, (forceProperty.nodes[i].x + forceProperty.nodes[i].width).toFloat(), (forceProperty.nodes[i].y + forceProperty.nodes[i].height + 30.0).toFloat(), paint)
                }

            }
            /** calculate spring-like forces */
            forceProperty.relax()
        }

        invalidate()
    }

    private fun midPoint(pointF: PointF, motionEvent: MotionEvent) {
        val x = motionEvent.getX(0) + motionEvent.getX(1)
        val y = motionEvent.getY(0) + motionEvent.getY(1)
        pointF.set(x / 2, y / 2)
    }
}